### ISSUE TITLE
Tests fail on NixOS - ModuleNotFoundError: No module named 'calmjs.types'

### ISSUE THREAD

**Issue by onny** [calmjs/calmjs#63]
Hi,
wenn trying to run pytest on calmjs I get following errors:
```
_______________ ERROR collecting src/calmjs/tests/test_vlqsm.py ________________
ImportError while importing test module '/build/calmjs-3.4.2/src/calmjs/tests/test_vlqsm.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/nix/store/rc9cz7z4qlgmsbwvpw2acig5g2rdws46-python3-3.10.5/lib/python3.10/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
src/calmjs/tests/test_vlqsm.py:7: in <module>
    from calmjs import vlqsm
src/calmjs/vlqsm.py:12: in <module>
    from calmjs.parse.vlq import (
E   ModuleNotFoundError: No module named 'calmjs.parse'
________________ ERROR collecting src/calmjs/tests/test_yarn.py ________________
ImportError while importing test module '/build/calmjs-3.4.2/src/calmjs/tests/test_yarn.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/nix/store/rc9cz7z4qlgmsbwvpw2acig5g2rdws46-python3-3.10.5/lib/python3.10/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
src/calmjs/tests/test_yarn.py:15: in <module>
    from calmjs import yarn
src/calmjs/yarn.py:14: in <module>
    from calmjs.runtime import PackageManagerRuntime
src/calmjs/runtime.py:28: in <module>
    from calmjs.artifact import ArtifactBuilder
src/calmjs/artifact.py:142: in <module>
    from calmjs.types.exceptions import ToolchainAbort
E   ModuleNotFoundError: No module named 'calmjs.types'
```

I'm trying to package the module for NixOS https://github.com/NixOS/nixpkgs/pull/186298/commits/87eff0c582a431049323be2c577119459797088d

As you can see there, calmjs-parse and calmjs-types are in `propagatedBuildInputs` meaning they are runtime dependencies. Inside the `checkPhase` both module paths are also present in PYTHONPATH.

Best regards
Jonas

**Comment by metatoaster** [comment]
Hi,

`PYTHONPATH` isn't exactly the "normal" way to do things in Python (even though it is possible to make it work); can you please output the `PYTHONPATH` environment variable that the Python interpreter sees during the test? I cannot exactly replicate the exact issue (although test will **fail** if the test is executed on the raw source without installing the package through the process understood by `setuptools`, explanation to follow) by doing the following:

```console
$ PYTHONPATH=src:../calmjs.parse/src:../calmjs.types/src python -m unittest calmjs.tests.make_suite
/build/calmjs/src/calmjs/cli.py:234: RuntimeWarning: Unable to locate the 'yarn' binary or runtime; default module level functions will not work. Please either provide PATH and/or update PATH environment variable with one that provides 'yarn'; or specify a working NODE_PATH environment variable with yarn installed; or have install 'yarn' into the current working directory (/build/calmjs) either through npm or calmjs framework for this package. Restart or reload this module once that is done. Alternatively, create a manual Driver instance for 'yarn' with explicitly defined arguments.
  warnings.warn(msg, RuntimeWarning)
encountered unexpected error while trying to find version of 'None':
Traceback (most recent call last):
  File "/build/calmjs/src/calmjs/cli.py", line 52, in get_bin_version_str
    prog = _get_exec_binary(bin_path, kw)
  File "/build/calmjs/src/calmjs/base.py", line 75, in _get_exec_binary
    binary = which(binary, path=kw.get('env', {}).get('PATH'))
  File "/build/calmjs/src/calmjs/utils.py", line 135, in which
    if os.path.dirname(cmd):
  File "/usr/lib/python-exec/python3.9/../../../lib/python3.9/posixpath.py", line 152, in dirname
    p = os.fspath(p)
TypeError: expected str, bytes or os.PathLike object, not NoneType
..............................................................................................................................................................FF.............................................................................................................................................................................................................................................................................................................................................................................................................................................................................................ss.ss...s.................sss.s.
======================================================================
FAIL: test_build_calmjs_artifact (test_dist.DistIntegrationTestCase)
Emulate the execution of ``python setup.py egg_info``.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/calmjs/src/calmjs/tests/test_dist.py", line 1171, in test_build_calmjs_artifact
    self.assertIn('running build_calmjs_artifacts', stdout)
AssertionError: 'running build_calmjs_artifacts' not found in ''

======================================================================
FAIL: test_setup_egg_info (test_dist.DistIntegrationTestCase)
Emulate the execution of ``python setup.py egg_info``.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/calmjs/src/calmjs/tests/test_dist.py", line 1131, in test_setup_egg_info
    self.assertIn('writing package_json', stdout)
AssertionError: 'writing package_json' not found in "running egg_info\ncreating dummy_pkg.egg-info\nwriting dummy_pkg.egg-info/PKG-INFO\nwriting dependency_links to dummy_pkg.egg-info/dependency_links.txt\nwriting top-level names to dummy_pkg.egg-info/top_level.txt\nwriting manifest file 'dummy_pkg.egg-info/SOURCES.txt'\nreading manifest file 'dummy_pkg.egg-info/SOURCES.txt'\nwriting manifest file 'dummy_pkg.egg-info/SOURCES.txt'\n"

----------------------------------------------------------------------
Ran 669 tests in 4.463s

FAILED (failures=2, skipped=9)
```

The two tests that fail is noted by the test for being able for `calmjs` to find the distribution (Python packaging) information, and without the relevant metadata written in a way that is understood by `setuptools` this test will fail, as this test is meant to be done against the fully installed and to verify that `calmjs` works after being installed to the active working Python environment.  Note that this is what might else that might happen if NixOS only does `PYTHONPATH` matching for Python packages, though I am sure this is not exactly the case.

If the requirement that this testing process (i.e. before preparation for installation) _must_ be done for NixOS, you may wish to apply set the environment variable `CALMJS_SKIP_INTEGRATION` to `1` (introduced as per #60), which will skip those two "problematic" tests for this particular situation. You may also wish to reference the discussion in #61 to see how OpenSUSE had to also fix a packaing issue in `calmjs.parse`, which may or may not also be affecting the way the setup harness is to be constructed to prepare for installation/packaging for NixOS.

Thank you for reporting this issue, and please do update me with the progress and/or the additional information should further assistance is required.

**Comment by metatoaster** [comment]
Closing as downstream has this item marked as merged and completed as per NixOS/nixpkgs#186298.
